synapse/federation/federation_client.py

```diff
@@ -495,16 +500,31 @@ async def _try_destination_list(
             except UnsupportedRoomVersionError:
                 raise
             except HttpResponseException as e:
-                if not 500 <= e.code < 600:
-                    raise e.to_synapse_error()
-                else:
-                    logger.warning(
-                        "Failed to %s via %s: %i %s",
-                        description,
-                        destination,
-                        e.code,
-                        e.args[0],
-                    )
+                synapse_error = e.to_synapse_error()
+                failover = False
+
+                if 500 <= e.code < 600:
+                    failover = True
+
+                elif failover_on_unknown_endpoint:
+                    # there is no good way to detect an "unknown" endpoint. Dendrite
+                    # returns a 404 (with no body); synapse returns a 400
+                    # with M_UNRECOGNISED.
+                    if e.code == 404 or (
+                        e.code == 400 and synapse_error.errcode == Codes.UNRECOGNIZED
+                    ):
+                        failover = True
+
+                if not failover:
```

Comment:
Should this be `if failover`? I'm having some trouble following this logic. Previously the `500` error would raise a `SynapseError`, but now it won't (since `failover` is `True` in that case).

Reply:
> Previously the 500 error would raise a SynapseError

I don't think that's right. Previously anything *except* a 500 error would raise a `SynapseError`.

I agree the logic here isn't super easy to follow. I'm very open to suggestions for how to reorganise it to make it clearer.

I was actually contemplating this logic in the shower this morning... Really, there ought to be a defined list of error codes that a well-behaved server can return for each endpoint - and anything else constitutes a misbehaving server whose response should be discounted. The trouble is that implementing that means getting into the weeds with refactoring the other code that uses `_try_destination_list`.

Reply:
Ah, yeah I had missed the `not` in front of the first clause. 😢 

I think the newer version is a bit clearer with the `failover` variable. I was just failing to see the equivalence to the old one.

I don't have any suggestions of how to make it better now. 👍